pkg/operation/botanist/component/vpnshoot/vpnshoot.go

```diff
@@ -0,0 +1,515 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vpnshoot
+
+import (
+	"context"
+	"time"
+
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	resourcesv1alpha1 "github.com/gardener/gardener/pkg/apis/resources/v1alpha1"
+	"github.com/gardener/gardener/pkg/client/kubernetes"
+	"github.com/gardener/gardener/pkg/operation/botanist/component"
+	"github.com/gardener/gardener/pkg/operation/botanist/component/vpnseedserver"
+	"github.com/gardener/gardener/pkg/resourcemanager/controller/garbagecollector/references"
+	kutil "github.com/gardener/gardener/pkg/utils/kubernetes"
+	"github.com/gardener/gardener/pkg/utils/managedresources"
+
+	appsv1 "k8s.io/api/apps/v1"
+	autoscalingv1 "k8s.io/api/autoscaling/v1"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	autoscalingv1beta2 "k8s.io/autoscaler/vertical-pod-autoscaler/pkg/apis/autoscaling.k8s.io/v1beta2"
+	"k8s.io/utils/pointer"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	// LabelValue is the value of a label used for the identification of vpn-shoot pods.
+	LabelValue = "vpn-shoot"
+	// ManagedResourceName is the name of the ManagedResource containing the resource specifications.
+	ManagedResourceName = "shoot-core-vpn-shoot"
+	// SecretNameTLSAuth is a constant for the tls auth secret name.
+	SecretNameTLSAuth = vpnseedserver.VpnSeedServerTLSAuth
+	// SecretNameDH is a constant for the dh secret name.
+	SecretNameDH = vpnseedserver.VpnSeedServerDH
+	// SecretName is the name of the shoot secret.
+	VpnShootSecretName = "vpn-shoot-client"
+
+	// servicePort is the service port used for the VPN LoadBalancer.
+	servicePort int32 = 4314
+	// containerPort is the target port used for the VPN LoadBalancer.
+	containerPort int32 = 1194
+	//deploymentName is the name of the vpnShoot deployment
+	deploymentName = "vpn-shoot"
+
+	containerName           = "vpn-shoot"
+	serviceName             = "vpn-test"
+	volumeNameSecret        = "vpn-shoot"
+	volumeNameSecretTLSAuth = "vpn-shoot-tlsauth"
+	volumeNameSecretDH      = "vpn-shoot-dh"
+
+	volumeMountPathSecret    = "/srv/secrets/vpn-shoot"
+	volumeMountPathSecretTLS = "/srv/secrets/tlsauth"
+	volumeMountPathSecretDH  = "/srv/secrets/dh"
+)
+
+// Interface contains functions for a VPNShoot Deployer
+type Interface interface {
+	component.DeployWaiter
+	component.MonitoringComponent
+	// SetSecrets sets the secrets.
+	SetSecrets(Secrets)
+}
+
+// Values is a set of configuration values for the VPNShoot component.
+type Values struct {
```

Comment:
We don't necessarily need a flat struct here, we can design this also better. What about introducing

```golang
type ReversedVPNValues struct {
  Enabled bool
  Header string
  Endpoint string
  OpenVPNPort string // btw - this should probably be int32
}
```

and 

```golang
type NetworkValues struct {
  PodCIDR string
  ServiceCIDR string
  NodeCIDR string
}
```
?

Reply:
Sounds like a good idea.

Reply:
I made Port as string because in the EnvVar struct, the value was string. yeah, but we can convert this to string and assign.